File: app/src/main/java/social/entourage/android/map/tour/information/discussion/LocationCardViewHolder.java
package social.entourage.android.map.tour.information.discussion;

import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;

import social.entourage.android.R;
import social.entourage.android.api.model.TimestampedObject;
import social.entourage.android.api.model.map.Tour;
import social.entourage.android.api.model.map.TourPoint;
import social.entourage.android.api.model.map.TourTimestamp;
import social.entourage.android.base.BaseCardViewHolder;

/**
 * Linear Layout that represents a location card in the tour info screen
 */
public class LocationCardViewHolder extends BaseCardViewHolder {

    private TextView mLocationDate;
    private ImageView mLocationImage;
    private TextView mLocationTitle;
    private TextView mLocationDuration;
    private TextView mLocationDistance;

    public LocationCardViewHolder(final View view) {
        super(view);
    }

    @Override
    protected void bindFields() {

        mLocationDate = (TextView) itemView.findViewById(R.id.tic_location_date);
        mLocationImage = (ImageView) itemView.findViewById(R.id.tic_location_image);
        mLocationTitle = (TextView) itemView.findViewById(R.id.tic_location_title);
        mLocationDuration = (TextView) itemView.findViewById(R.id.tic_location_duration);
        mLocationDistance = (TextView) itemView.findViewById(R.id.tic_location_distance);
    }

    @Override
    public void populate(final TimestampedObject data) {
        populate((TourTimestamp) data);
    }

    public void populate(Tour tour, boolean isStartCard) {

        List<TourPoint> tourPointsList = tour.getTourPoints();

        SimpleDateFormat locationDateFormat = new SimpleDateFormat(itemView.getResources().getString(R.string.tour_info_location_card_date_format));
        if (!isStartCard) {
            mLocationDate.setText(locationDateFormat.format(tour.getEndTime()));
            mLocationTitle.setText(R.string.tour_info_text_closed);
            if (tour.isClosed()) {
                if (tour.getStartTime() != null && tour.getEndTime() != null) {
                    Date duration = new Date(tour.getEndTime().getTime() - tour.getStartTime().getTime());
                    SimpleDateFormat dateFormat = new SimpleDateFormat("HH:mm:ss", Locale.US);
                    dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
                    mLocationDuration.setText(dateFormat.format(duration));
                }
                float distance = 0;
                TourPoint startPoint = tourPointsList.get(0);
                for (int i=1; i < tourPointsList.size(); i++) {
                    TourPoint p = tourPointsList.get(i);
                    distance += p.distanceTo(startPoint);
                    startPoint = p;
                }
                mLocationDistance.setText(String.format("%.2f km", distance/1000.0f));
            }
        } else {
            mLocationDate.setText(locationDateFormat.format(tour.getStartTime()));
            mLocationTitle.setText(R.string.tour_info_text_ongoing);
            if (!tour.isClosed() && tour.getStartTime() != null) {
                Date duration = new Date((new Date()).getTime() - tour.getStartTime().getTime());
                SimpleDateFormat dateFormat = new SimpleDateFormat("HH:mm:ss", Locale.US);
                dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
                mLocationDuration.setText(dateFormat.format(duration));
            }
            mLocationDistance.setText("");
        }

    }

    public void populate(TourTimestamp tourTimestamp) {
        SimpleDateFormat locationDateFormat = new SimpleDateFormat(itemView.getResources().getString(R.string.tour_info_location_card_date_format));
        mLocationDate.setText(locationDateFormat.format(tourTimestamp.getDate()));

        if (Tour.TOUR_ON_GOING.equals(tourTimestamp.getStatus())) {
            mLocationTitle.setText(R.string.tour_info_text_ongoing);
        }
        else {
            mLocationTitle.setText(R.string.tour_info_text_closed);
        }

        if (tourTimestamp.getDistance() > 0) {
            mLocationDistance.setText(String.format("%.2f km", tourTimestamp.getDistance()/1000.0f));
            mLocationDistance.setVisibility(View.VISIBLE);
        }
        else {
            mLocationDistance.setVisibility(View.GONE);
        }

        if (tourTimestamp.getDuration() > 0) {
            Date duration = new Date(tourTimestamp.getDuration());
            SimpleDateFormat dateFormat = new SimpleDateFormat("HH:mm:ss", Locale.US);
            dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
            mLocationDuration.setText(dateFormat.format(duration));
            mLocationDuration.setVisibility(View.VISIBLE);
        }
        else {
            mLocationDuration.setVisibility(View.GONE);
        }

        if (tourTimestamp.getSnapshot() != null) {
            mLocationImage.setImageBitmap(tourTimestamp.getSnapshot());
            mLocationImage.setVisibility(View.VISIBLE);
        }
        else {
            mLocationImage.setVisibility(View.GONE);
        }
    }

    public static int getLayoutResource() {
        return R.layout.tour_information_location_card_view;
    }

}

File: app/src/main/java/social/entourage/android/entourage/information/discussion/LocationCardViewHolder.kt
package social.entourage.android.entourage.information.discussion

import android.view.View
import kotlinx.android.synthetic.main.layout_tour_information_location_card_view.view.*
import social.entourage.android.R
import social.entourage.android.api.model.TimestampedObject
import social.entourage.android.api.model.feed.FeedItem
import social.entourage.android.api.model.tour.Tour
import social.entourage.android.api.model.tour.TourInformation
import social.entourage.android.base.BaseCardViewHolder
import social.entourage.android.tools.Utils
import java.text.SimpleDateFormat
import java.util.*

/**
 * Linear Layout that represents a location card in the tour info screen
 */
class LocationCardViewHolder(view: View) : BaseCardViewHolder(view) {

    override fun bindFields() {}

    override fun populate(data: TimestampedObject) {
        populate(data as TourInformation)
    }

    private fun populate(tour: Tour, isStartCard: Boolean) {
        val tourPointsList = tour.tourPoints
        val locationDateFormat = SimpleDateFormat(itemView.resources.getString(R.string.tour_info_location_card_date_format), Locale.FRANCE)
        if (!isStartCard) {
            itemView.tic_location_date?.text = locationDateFormat.format(tour.getEndTime() ?: Date()).toUpperCase(Locale.ROOT)
            itemView.tic_location_title?.setText(R.string.tour_info_text_closed)
            if (tour.isClosed()) {
                tour.getEndTime()?.let {endTime->
                    itemView.tic_location_duration?.text = Utils.getDateStringFromSeconds(endTime.time - tour.getStartTime().time)
                }
                var distance = 0f
                var startPoint = tourPointsList[0]
                for (i in 1 until tourPointsList.size) {
                    val p = tourPointsList[i]
                    distance += p.distanceTo(startPoint)
                    startPoint = p
                }
                itemView.tic_location_distance?.text = String.format("%.2f km", distance / 1000.0f)
            }
        } else {
            itemView.tic_location_date?.text = locationDateFormat.format(tour.getStartTime()).toUpperCase(Locale.ROOT)
            itemView.tic_location_title?.setText(R.string.tour_info_text_ongoing)
            if (!tour.isClosed()) {
                itemView.tic_location_duration?.text = Utils.getDateStringFromSeconds(Date().time - tour.getStartTime().time)
            }
            itemView.tic_location_distance?.text = ""
        }
    }

    fun populate(tourInformation: TourInformation) {
        val locationDateFormat = SimpleDateFormat(itemView.resources.getString(R.string.tour_info_location_card_date_format), Locale.FRANCE)
        itemView.tic_location_date?.text = locationDateFormat.format(tourInformation.startDate).toUpperCase(Locale.ROOT)
        if (FeedItem.STATUS_ON_GOING == tourInformation.status || FeedItem.STATUS_OPEN == tourInformation.status) {
            if (tourInformation.feedType == TimestampedObject.TOUR_CARD) {
                itemView.tic_location_title?.setText(R.string.tour_info_text_ongoing)
            } else if (tourInformation.feedType == TimestampedObject.ENTOURAGE_CARD) {
                itemView.tic_location_title?.setText(R.string.entourage_info_text_open)
            }
        } else {
            if (tourInformation.feedType == TimestampedObject.TOUR_CARD) {
                itemView.tic_location_title?.setText(R.string.tour_info_text_closed)
            } else if (tourInformation.feedType == TimestampedObject.ENTOURAGE_CARD) {
                itemView.tic_location_title?.setText(R.string.entourage_info_text_close)
            }
        }
        if (tourInformation.distance > 0) {
            itemView.tic_location_distance?.text = String.format("%.2f km", tourInformation.distance / 1000.0f)
            itemView.tic_location_distance?.visibility = View.VISIBLE
        } else {
            itemView.tic_location_distance?.visibility = View.GONE
        }
        if (tourInformation.duration > 0) {
            itemView.tic_location_duration?.text = Utils.getDateStringFromSeconds(tourInformation.duration)
            itemView.tic_location_duration?.visibility = View.VISIBLE
        } else {
            itemView.tic_location_duration?.visibility = View.GONE
        }
        if (tourInformation.snapshot != null) {
            itemView.tic_location_image?.setImageBitmap(tourInformation.snapshot)
            itemView.tic_location_image?.visibility = View.VISIBLE
        } else {
            itemView.tic_location_image?.visibility = View.GONE
        }
    }

    companion object {
        @JvmStatic
        val layoutResource: Int
            get() = R.layout.layout_tour_information_location_card_view
    }
}